# immuVCF

A whole exome sequence (WES) bioinformatics pipeline for annotating immune-case variant call files (VCFs).

## Background

Coming soon!

## Pipeline Overview

The pipeline is run on Python 3. See dependencies for necessary packages.

The main mode of the pipeline is ‘annotate_VCF’, which generates annotations and exports the updated VCF. 

Input Files (See ‘Input Data’ for more information):
(1) : Whole-Exome Variant Call File
(2) : HPO Terms

| **Mode**              | **Purpose**                                                                                                                                                                                                                                                                                           | **Input Dependencies** |
|-----------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------|
| ‘annotate_VCF’        | The main mode of the pipeline. Generates annotations and exports the updated VCF.                                                                                                                                                                                                                     | (1) and (2)            |
| ‘generate_STRING_df’  | Generates a CSV dataframe file necessary to run ‘annotate_VCF’ containing data from the STRING database. This mode must be run every time STRING data is updated.  Note: Downloading the files from this repository will automatically contain the updated dataframe, and this mode is not necessary. | None                   |
| ‘generate_BioGRID_df’ | Similar to above, except data from the BioGRID database. Again, this mode must be run every time BioGRID data is updated.  Note: Downloading the files from this repository will automatically contain the updated dataframe, and this mode is not necessary.                                         | None                   |
| ‘annotate_iei_list’   | Generates and exports a list of annotated genes known to cause IEI.                                                                                                                                                                                                                                   | None                   |
| ‘generate_HPO_list’   | Generates and exports a list of genes that appear in the HPO analysis.                                                                                                                                                                                                                                | (2)                    |

### Input Data: 

Both should be directed into ‘input_files’

(1) Whole-Exome Variant Call File (this can be annotated already by another pipeline or not). Excel (.xlsx) format. If there are multiple tabs in the excel sheet, it must be the first tab. A column with the Ensembl Gene ID is recommended but not necessary.

(2) HPO Terms: Two accepted input formats
	1	HPO ID List: A text file with a plain list of HPO identification numbers. No labels, extra blank lines, or whitespace. One ID per line.
	2	Phenotips Export: An exported .tsv containing HPO genes and their related HPO terms

### Configurations

Pipeline run settings can be altered in ‘config.py’. Below is an explanation each configuration:

| **Parameter**                                        | **Acceptable Inputs**                                                                                       | **Explanation**                                                                                                                                                                                                                                                                      |
|------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| mode                                                 | ‘annotate_VCF’,  ‘generate_STRING_df’, ‘generate_BioGRID_df’, ‘generate_HPO_list’, ‘annotate_iei_list’      | What mode to run. See above table for explanations of each mode.                                                                                                                                                                                                                     |
| vcf_filename                                         | Path to the VCF in Excel (.xlsx) format.                                                                    | VCF Input                                                                                                                                                                                                                                                                            |
| hpo_id_filename                                      | Path to the HPO file.                                                                                       | HPO Input                                                                                                                                                                                                                                                                            |
| write_to                                             | Path to the output file                                                                                     | Writes VCF or gene list to this location.  **Must end in .xlsx!**                                                                                                                                                                                                                                           |
| generate_genes                                       | True or False                                                                                               | True: ‘annotate_VCF’ will annotate the VCF with a column containing the gene name the variant is on. Requires the parameter ‘coord_column’. False: ‘annotate_VCF’ will generate all annotations based on an existing Ensembl Gene ID column. Requires the parameter ‘ensembl_column’ |
| input_hpo                                            | 'Phenotypes' or 'List'                                                                                      | ‘Phenotips’: Imports HPO terms using the specified Phenotips format. ‘List’: Imports HPO terms using the specified HPO format.                                                                                                                                                       |
| separate_immune_hpo_terms                            | True or False                                                                                               | True: Separates all HPO annotations based on whether the HPO term relates to immunity. False: Groups all HPO terms under the same column.                                                                                                                                            |
| filter_non_protein_genes                             | True or False                                                                                               | Only when ‘generate_genes’ is True.  True: Only shows generated genes that are protein coding.                                                                                                                                                                                       |
| string_threshold                                     | Integer from 1-1000                                                                                         | Filters all STRING interactions not above the threshold value. The threshold is usually set to ‘medium’ or 700.                                                                                                                                                                      |
| reference_genome                                     | ‘GRCh37’ or ‘GRCh38’                                                                                        | Only when ‘generate_genes’ is True. Which human reference genome to use for gene locations.                                                                                                                                                                                          |
| Gene ID Files, HPO Files, Interaction Analysis Files | Path to the respective database file.                                                                       | Database file downloads. See below table.                                                                                                                                                                                                                                            |
| Generated Dataframe                                  | Output for ‘generate_STRING_df’ or ‘generate_BioGRID_df’. Input for ‘annotate_VCF’ and ‘annotate_IEI_list’. | Path to generated interaction analysis dataframe.                                                                                                                                                                                                                                    |
| ensembl_column                                       | Name of the Ensembl Gene ID column on the input VCF                                                         | Only when ‘generate_genes’ is False                                                                                                                                                                                                                                                  |
| coord_column                                         | Name of the “chromosome:coordinate” position column on the input VCF.                                       | Only when ‘generate_genes’ is True                                                                                                                                                                                                                                                   |
| duplicate_column_label                               | What to name the duplicate column label.                                                                    | Only when ‘generate_genes’ is True                                                                                                                                                                                                                                                   |
| gene_column_label                                    | What to name the generated gene name column label.                                                          | Only when ‘generate_genes’ is True                                                                                                                                                                                                                                                   |

## Dependencies
Packages: The pipeline depends on the latest versions of ‘pandas’ and ‘openpyxl’. Both packages must be installed before use.

### Data Files

Stored in ‘file_dependencies’

| **Name**                                                      | **Current File Handle**                | **Notes**                                                                                                                                                                                                                                                                             | **Source**                                                                                                                                                                                                                |
|---------------------------------------------------------------|----------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Known Gene List iei_gene_filename                             | ‘IEI_Genes_{date_updated}.xlsx’        | Known IEI disease-causing genes. Clear formatting and no column labels. 3 Tabs. First tab is all IEI genes. Second tab is all IEI genes on the Blueprint Genetics Panel List. Third tab is all IEI genes on the VEO-IBD gene list. (All lists can be provided by Dr. Ashish Marwaha). | Dr. Marwaha - Alberta Children’s Hospital                                                                                                                                                                                 |
| HGNC Gene Identifications hgnc_filename                       | ‘hgnc_gene_id.txt’                     | Different gene identifications compiled by HGNC.                                                                                                                                                                                                                                      | From HGNC custom downloads:  https://www.genenames.org/download/custom/Select HGNC ID, Approved Symbol, NCBI Gene ID, Ensembl Gene ID, RefSeq IDs.                                                                        |
| Ensembl Protein and Genes gp_filename                         | ‘gene_proteins.txt’                    | Ensembl gene and protein ID relationships.                                                                                                                                                                                                                                            | From BioMart: https://uswest.ensembl.org/info/data/biomart/index.htmlSelect human dataset (reference should not matter), filter by protein coding genes. Select Ensembl Gene ID and Protein ID attributes. Export as tsv. |
| STRING Protein Interactions string_filename                   | ‘STRING_data.txt’                      | STRING interactions and compiled scores.                                                                                                                                                                                                                                              | From STRING: https://string-db.org/cgi/download?sessionId=bJbN9ZRf6A6ODownload the first general links file after specifying Homo-sapiens.                                                                                |
| HPO Annotations ph2g_filename                                 | ‘phenotype_to_genes.txt’               | HPO Phenotype terms and annotations.                                                                                                                                                                                                                                                  | From HPO: https://hpo.jax.org/app/download/annotationDownload phenotype to genes.                                                                                                                                         |
| HPO Annotations g2ph_filename                                 | ‘genes_to_phenotype.txt’               | HPO genes and annotations                                                                                                                                                                                                                                                             | From HPO: https://hpo.jax.org/app/download/annotationDownload genes to phenotype.                                                                                                                                         |
| BIOGRID Interactions bio_filename                             | ‘BIOGRID_data.txt’                     | All BIOGRID Interactions                                                                                                                                                                                                                                                              | From BIOGRID: https://downloads.thebiogrid.org/BioGRIDSelect current release, ORGANISM-tab3. Unzip and select the Homo_sapiens file.                                                                                      |
| Closest Disease Causing Gene List cdg_filename                | ‘CDG_data.txt’                         | Closest associated genes to known IEI genes. **THIS DOCUMENT REQUIRED RE-FORMATTING OF THE HEADERS:Gene<tab>HGMD_Gene<tab>HPO<tab>P_Value<tab>Route<tab>Associated_Disease**                                                                                                              | From CDG server (based on HGMD): https://lab.rockefeller.edu/casanova/CDGPaste in all genes from known gene list. Click download results on the top of page.Make sure to reformat headers (see notes).                    |
| GENCODE GRCh38 and GRCh37 annotations gencode_GRCh3X_filename | ‘GENCODE_GRCh3X.gtf’                   | Chromosome and base pair coordinates for all human genes.                                                                                                                                                                                                                             | From Gencode: https://www.gencodegenes.orgDownload both GRCh38 and GRCh37 (in GTF format)                                                                                                                                 |
| Immune HPO Terms immune_HPO_filename                          | ‘immune_HPO_terms_{date_updated}.xlsx’ | Static Document Includes all relevant immune-related HPO terms in CSV format                                                                                                                                                                                                          | Created by Benjamin Green.                                                                                                                                                                                                |

## Instructions to Run

Youtube video tutorials coming soon!

(1) Ensure the pipeline’s files and file dependencies are all downloaded and in the correct directories

(2) Place the VCF (in .xlsx) and HPO (.tsv) files inside the folder ‘input_dependencies’

(3) Open ‘config.py’ with TextEdit or another text editor

(4) Make sure the parameter ‘mode’ is set to ‘annotate_VCF’. Adjust the parameters ‘vcf_filename’ and ‘hpo_id_filename’ so that they point to the relevant files.

(5) Adjust the parameter ‘write_to’ to where you want to write the annotated VCF. Typically, this is in the ‘generated_files’ directory.

(6) Save, and then close the file.

(7) On Mac: Open the terminal app. Change the directory to the ‘immuVCF’  folder using the command ‘cd immuVCF’. If you are lost, use the command ‘ls’ to list all files and directories in your current directory.

(8) Once inside the folder ‘immuVCF’, run ‘main.py’ on Python by running the command: ‘python3 main.py’


